File: go/ch21/mock_contest.go
// go/ch21/mock_contest.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"math"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
)

type ProblemInput struct {
	Days  int
	Decay []int
	Gain  [][]int
}

type ScheduleState struct {
	Types []int
}

type SubmissionLog struct {
	Phase      string  `json:"phase"`
	ElapsedSec float64 `json:"elapsed_sec"`
	Score      int     `json:"score"`
	Note       string  `json:"note"`
}

func readProblem(path string) (ProblemInput, error) {
	var f *os.File
	var err error
	if path == "" {
		f = os.Stdin
	} else {
		f, err = os.Open(path)
		if err != nil {
			return ProblemInput{}, err
		}
		defer f.Close()
	}
	sc := bufio.NewScanner(f)
	var lines []string
	for sc.Scan() {
		lines = append(lines, sc.Text())
	}
	if err := sc.Err(); err != nil {
		return ProblemInput{}, err
	}
	days, _ := strconv.Atoi(lines[0])
	decay := parseInts(lines[1])
	gain := make([][]int, days)
	for i := 0; i < days; i++ {
		gain[i] = parseInts(lines[2+i])
	}
	return ProblemInput{days, decay, gain}, nil
}

func parseInts(line string) []int {
	parts := strings.Fields(line)
	out := make([]int, len(parts))
	for i, p := range parts {
		out[i], _ = strconv.Atoi(p)
	}
	return out
}

func scoreSchedule(p ProblemInput, s ScheduleState) int {
	last := make([]int, 26)
	satisfaction := 0
	for day := 1; day <= p.Days; day++ {
		t := s.Types[day-1] - 1
		satisfaction += p.Gain[day-1][t]
		last[t] = day
		decay := 0
		for i := 0; i < 26; i++ {
			decay += p.Decay[i] * (day - last[i])
		}
		satisfaction -= decay
	}
	score := 1000000 + satisfaction
	if score < 0 {
		return 0
	}
	return score
}

func finalSat(p ProblemInput, types []int) int {
	last := make([]int, 26)
	satisfaction := 0
	for day := 1; day <= len(types); day++ {
		t := types[day-1] - 1
		satisfaction += p.Gain[day-1][t]
		last[t] = day
		decay := 0
		for i := 0; i < 26; i++ {
			decay += p.Decay[i] * (day - last[i])
		}
		satisfaction -= decay
	}
	return satisfaction
}

func greedySchedule(p ProblemInput) ScheduleState {
	schedule := make([]int, 0, p.Days)
	for day := 1; day <= p.Days; day++ {
		bestT, bestSat := 1, -1<<30
		for t := 1; t <= 26; t++ {
			trial := append(append([]int{}, schedule...), t)
			sat := finalSat(p, trial)
			if sat > bestSat {
				bestSat = sat
				bestT = t
			}
		}
		schedule = append(schedule, bestT)
	}
	return ScheduleState{schedule}
}

func briefSA(p ProblemInput, state ScheduleState, deadline time.Time, rng *rand.Rand) ScheduleState {
	current := append([]int{}, state.Types...)
	score := scoreSchedule(p, ScheduleState{current})
	best := append([]int{}, current...)
	bestScore := score
	temperature := 5000.0
	alpha := 0.995

	for time.Now().Before(deadline) {
		day := rng.Intn(p.Days)
		trial := append([]int{}, current...)
		trial[day] = 1 + rng.Intn(26)
		trialScore := scoreSchedule(p, ScheduleState{trial})
		delta := trialScore - score
		if delta >= 0 || rng.Float64() < math.Exp(float64(delta)/temperature) {
			current = trial
			score = trialScore
			if score > bestScore {
				best = append([]int{}, current...)
				bestScore = score
			}
		}
		temperature *= alpha
	}
	return ScheduleState{best}
}

type MockContest struct {
	problem   ProblemInput
	phaseMin  []float64
	fast      bool
	jsonLog   bool
	start     time.Time
	logs      []SubmissionLog
	rng       *rand.Rand
	state     ScheduleState
	bestScore int
}

func (c *MockContest) elapsed() float64 {
	return time.Since(c.start).Seconds()
}

func (c *MockContest) emit(phase, event string, score int, note string) {
	entry := SubmissionLog{
		Phase:      phase + ":" + event,
		ElapsedSec: c.elapsed(),
		Score:      score,
		Note:       note,
	}
	c.logs = append(c.logs, entry)
	fmt.Fprintf(os.Stderr, "[contest] phase=%s event=%s t=%.2fs score=%d %s\n",
		phase, event, entry.ElapsedSec, score, note)
	if c.jsonLog {
		b, _ := json.Marshal(entry)
		fmt.Fprintln(os.Stderr, string(b))
	}
}

func (c *MockContest) waitPhase(name string, minutes float64) {
	sec := minutes * 60.0
	if c.fast {
		sec = minutes
	}
	end := c.elapsed() + sec
	c.emit(name, "start", c.bestScore, fmt.Sprintf("duration=%.1fs", sec))
	for c.elapsed() < end {
		time.Sleep(50 * time.Millisecond)
	}
	c.emit(name, "end", c.bestScore, "phase complete")
}

func (c *MockContest) run() ScheduleState {
	c.waitPhase("read", c.phaseMin[0])
	c.waitPhase("implement", c.phaseMin[1])
	c.emit("implement", "work", c.bestScore, "greedy build")
	c.state = greedySchedule(c.problem)
	c.bestScore = scoreSchedule(c.problem, c.state)
	c.emit("implement", "submit", c.bestScore, "initial solution")

	c.waitPhase("improve", c.phaseMin[2])
	improveDur := c.phaseMin[2] * 60.0
	if c.fast {
		improveDur = c.phaseMin[2]
	}
	deadline := time.Now().Add(time.Duration(improveDur * float64(time.Second)))
	c.emit("improve", "work", c.bestScore, "brief SA")
	c.state = briefSA(c.problem, c.state, deadline, c.rng)
	c.bestScore = scoreSchedule(c.problem, c.state)
	c.emit("improve", "submit", c.bestScore, "improved solution")

	c.waitPhase("submit", c.phaseMin[3])
	c.emit("submit", "final", c.bestScore, "stdout schedule")
	return c.state
}

func defaultInput() string {
	for _, rel := range []string{"data/ch23/sample.txt", "data/ch08/sample.txt"} {
		if _, err := os.Stat(rel); err == nil {
			return rel
		}
	}
	return "data/ch23/sample.txt"
}

func main() {
	fast := false
	jsonLog := false
	inputPath := ""
	for _, arg := range os.Args[1:] {
		switch arg {
		case "--fast":
			fast = true
		case "--json-log":
			jsonLog = true
		default:
			if !strings.HasPrefix(arg, "-") {
				inputPath = arg
			}
		}
	}
	if os.Getenv("CONTEST_FAST") == "1" {
		fast = true
	}
	if inputPath == "" {
		inputPath = defaultInput()
		fmt.Fprintf(os.Stderr, "input=%s\n", inputPath)
	}

	problem, err := readProblem(inputPath)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}

	phaseMin := []float64{30, 60, 90, 30}
	if fast {
		phaseMin = []float64{0.3, 0.5, 0.8, 0.2}
	}
	types := make([]int, problem.Days)
	for i := range types {
		types[i] = 1
	}
	contest := &MockContest{
		problem:   problem,
		phaseMin:  phaseMin,
		fast:      fast,
		jsonLog:   jsonLog,
		start:     time.Now(),
		rng:       rand.New(rand.NewSource(42)),
		state:     ScheduleState{types},
		bestScore: scoreSchedule(problem, ScheduleState{types}),
	}
	final := contest.run()
	for _, t := range final.Types {
		fmt.Println(t)
	}
	fmt.Fprintf(os.Stderr, "final_score=%d elapsed=%.2fs submissions=%d\n",
		contest.bestScore, contest.elapsed(), len(contest.logs))
}

File: python/ch21/mock_contest.py
# python/ch21/mock_contest.py
"""模擬 AHC: 読解 → 実装 → 改善 → 提出の 4 フェーズ（タイマー・提出ログ）."""
from __future__ import annotations

import argparse
import json
import math
import random
import sys
import time
from dataclasses import asdict, dataclass
from pathlib import Path
from typing import List, Optional, TextIO


@dataclass
class ProblemInput:
    days: int
    decay: List[int]
    gain: List[List[int]]


@dataclass
class ScheduleState:
    types: List[int]


@dataclass
class SubmissionLog:
    phase: str
    elapsed_sec: float
    score: int
    note: str


def read_problem(stream: TextIO) -> ProblemInput:
    lines = stream.read().splitlines()
    idx = 0
    days = int(lines[idx])
    idx += 1
    decay = list(map(int, lines[idx].split()))
    idx += 1
    gain = [list(map(int, lines[idx + i].split())) for i in range(days)]
    return ProblemInput(days=days, decay=decay, gain=gain)


def score_schedule(problem: ProblemInput, state: ScheduleState) -> int:
    last = [0] * 26
    satisfaction = 0
    for day in range(1, problem.days + 1):
        t = state.types[day - 1] - 1
        satisfaction += problem.gain[day - 1][t]
        last[t] = day
        decay = sum(problem.decay[i] * (day - last[i]) for i in range(26))
        satisfaction -= decay
    return max(10**6 + satisfaction, 0)


def greedy_schedule(problem: ProblemInput) -> ScheduleState:
    schedule: List[int] = []
    for day in range(1, problem.days + 1):
        best_t, best_sat = 1, None
        for t in range(1, 27):
            trial = ScheduleState(schedule + [t])
            sat = _final_satisfaction(problem, trial)
            if best_sat is None or sat > best_sat:
                best_sat = sat
                best_t = t
        schedule.append(best_t)
    return ScheduleState(schedule)


def _final_satisfaction(problem: ProblemInput, state: ScheduleState) -> int:
    last = [0] * 26
    satisfaction = 0
    for day in range(1, len(state.types) + 1):
        t = state.types[day - 1] - 1
        satisfaction += problem.gain[day - 1][t]
        last[t] = day
        decay = sum(problem.decay[i] * (day - last[i]) for i in range(26))
        satisfaction -= decay
    return satisfaction


def brief_sa(
    problem: ProblemInput,
    state: ScheduleState,
    deadline: float,
    rng: random.Random,
) -> ScheduleState:
    current = ScheduleState(list(state.types))
    score = score_schedule(problem, current)
    best = ScheduleState(list(current.types))
    best_score = score
    temperature = 5000.0
    alpha = 0.995

    while time.perf_counter() < deadline:
        day = rng.randrange(problem.days)
        trial = ScheduleState(list(current.types))
        trial.types[day] = rng.randint(1, 26)
        trial_score = score_schedule(problem, trial)
        delta = trial_score - score
        if delta >= 0 or rng.random() < math.exp(delta / temperature):
            current = trial
            score = trial_score
            if score > best_score:
                best = ScheduleState(list(current.types))
                best_score = score
        temperature *= alpha
    return best


class MockContest:
    def __init__(
        self,
        problem: ProblemInput,
        phase_min: List[float],
        fast: bool,
        log_json: bool,
    ) -> None:
        self.problem = problem
        self.phase_min = phase_min
        self.fast = fast
        self.log_json = log_json
        self.start = time.perf_counter()
        self.logs: List[SubmissionLog] = []
        self.rng = random.Random(42)
        self.state = ScheduleState([1] * problem.days)
        self.best_score = score_schedule(problem, self.state)

    def elapsed(self) -> float:
        return time.perf_counter() - self.start

    def wait_phase(self, name: str, minutes: float) -> None:
        sec = minutes * 60.0 if not self.fast else minutes
        end = self.elapsed() + sec
        self._emit(name, "start", self.best_score, f"duration={sec:.1f}s")
        while self.elapsed() < end:
            time.sleep(min(0.05, end - self.elapsed()))
        self._emit(name, "end", self.best_score, "phase complete")

    def _emit(self, phase: str, event: str, score: int, note: str) -> None:
        entry = SubmissionLog(
            phase=f"{phase}:{event}",
            elapsed_sec=self.elapsed(),
            score=score,
            note=note,
        )
        self.logs.append(entry)
        line = f"[contest] phase={phase} event={event} t={entry.elapsed_sec:.2f}s score={score} {note}"
        print(line, file=sys.stderr)
        if self.log_json:
            print(json.dumps(asdict(entry), ensure_ascii=False), file=sys.stderr)

    def run(self) -> ScheduleState:
        self.wait_phase("read", self.phase_min[0])
        self.wait_phase("implement", self.phase_min[1])
        self._emit("implement", "work", self.best_score, "greedy build")
        self.state = greedy_schedule(self.problem)
        self.best_score = score_schedule(self.problem, self.state)
        self._emit("implement", "submit", self.best_score, "initial solution")

        self.wait_phase("improve", self.phase_min[2])
        improve_end = self.elapsed() + (
            self.phase_min[2] * 60.0 if not self.fast else self.phase_min[2]
        )
        self._emit("improve", "work", self.best_score, "brief SA")
        self.state = brief_sa(self.problem, self.state, improve_end, self.rng)
        self.best_score = score_schedule(self.problem, self.state)
        self._emit("improve", "submit", self.best_score, "improved solution")

        self.wait_phase("submit", self.phase_min[3])
        self._emit("submit", "final", self.best_score, "stdout schedule")
        return self.state


def default_input_path() -> Path:
    root = Path(__file__).resolve().parents[2]
    for rel in ("data/ch23/sample.txt", "data/ch08/sample.txt"):
        p = root / rel
        if p.exists():
            return p
    return root / "data/ch23/sample.txt"


def main() -> None:
    parser = argparse.ArgumentParser(description="模擬 AHC 4 フェーズ")
    parser.add_argument(
        "input",
        nargs="?",
        help="Intro HC 形式の入力（省略時は data/ch23/sample.txt）",
    )
    parser.add_argument(
        "--fast",
        action="store_true",
        help="短縮フェーズ（秒単位: 0.3,0.5,0.8,0.2）",
    )
    parser.add_argument("--json-log", action="store_true", help="提出ログを JSON 行でも出力")
    args = parser.parse_args()

    fast = args.fast or __import__("os").environ.get("CONTEST_FAST") == "1"
    if args.input == "-":
        problem = read_problem(sys.stdin)
    elif args.input:
        with open(args.input, encoding="utf-8") as f:
            problem = read_problem(f)
    else:
        path = default_input_path()
        with path.open(encoding="utf-8") as f:
            problem = read_problem(f)
        print(f"input={path}", file=sys.stderr)

    phase_min = [0.3, 0.5, 0.8, 0.2] if fast else [30.0, 60.0, 90.0, 30.0]
    contest = MockContest(problem, phase_min, fast=fast, log_json=args.json_log)
    final = contest.run()
    for t in final.types:
        print(t)
    print(
        f"final_score={contest.best_score} elapsed={contest.elapsed():.2f}s submissions={len(contest.logs)}",
        file=sys.stderr,
    )


if __name__ == "__main__":
    main()
